File: 3.services/src/service_py_pkg/service_py_pkg/acc_server.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

from custom_interfaces.srv import Acc


class AccServer(Node):
    """A simple ROS2 node that creates a server for the service "accumulate" and accumulates the received numbers."""

    def __init__(self):
        super().__init__("acc_server_node")
        self.get_logger().info("Accumulate Server Node has been started!")
        # Create a service
        self.srv = self.create_service(Acc, "accumulate", self.accumulate_callback)

    def accumulate_callback(self, request: Acc.Request, response: Acc.Response):
        """Callback function for the service "accumulate"."""
        self.get_logger().info(
            f"Incoming request: a={request.a}, b={request.b}, c={request.c}"
        )
        response.sum = request.a + request.b + request.c
        return response


def main(args=None):
    # Initialize the ROS2 Python client library
    rclpy.init(args=args)

    # Create a ROS2 node, then spin the node to keep it alive, finally destroy the node
    node = AccServer()
    rclpy.spin(node)
    node.destroy_node()

    # Shutdown the ROS2 Python client library
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: 3.services/src/service_cpp_pkg/src/acc_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "custom_interfaces/srv/acc.hpp"

class AccServerNode : public rclcpp::Node
{
public:
    AccServerNode() : Node("acc_server_node")
    {
        // Create a service
        service_ = this->create_service<custom_interfaces::srv::Acc>(
            "accumulate",
            std::bind(&AccServerNode::handle_accumulate, this, std::placeholders::_1, std::placeholders::_2));
    }

private:
    void handle_accumulate(
        const std::shared_ptr<custom_interfaces::srv::Acc::Request> request,
        std::shared_ptr<custom_interfaces::srv::Acc::Response> response)
    {
        response->sum = request->a + request->b + request->c;
        RCLCPP_INFO(this->get_logger(), "Incoming request: a=%ld, b=%ld, c=%ld", request->a, request->b, request->c);
    }

    rclcpp::Service<custom_interfaces::srv::Acc>::SharedPtr service_;
};

int main(int argc, char *argv[])
{
    // Initialize the ROS 2 client library
    rclcpp::init(argc, argv);

    // Create a custom node
    auto node = std::make_shared<AccServerNode>();
    // Keep the node alive until it is shut down
    rclcpp::spin(node);
    // Destroy the node
    node.reset();

    // Shutdown the ROS 2 client library
    rclcpp::shutdown();
    return 0;
}
